### PR TITLE
Fix unsanitized nulls produced by libcudf dictionary decode

### PR DESCRIPTION
**Pull request by davidwendt**
## Description
Fixes `cudf::dictionary::decode` logic to produced sanitized null entries for compound column types.

Reference: #14559 -- fixes many of the errors found here concerning dictionary column gtests.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/src/dictionary/decode.cu [hunk 1]
```diff
@@ -30,6 +29,18 @@
 namespace cudf {
 namespace dictionary {
 namespace detail {
+namespace {
+struct indices_handler_fn {
+  cudf::detail::input_indexalator const d_iterator;
+  column_device_view const d_indices;
+  size_type oob_index;  // out-of-bounds index identifies nulls
+  __device__ size_type operator()(size_type idx)
```
Comment (ttnghia) [inline cpp/src/dictionary/decode.cu:37]:
```suggestion
  __device__ size_type operator()(size_type idx) const
```

**Review by ttnghia — APPROVED** [review]

File: cpp/src/dictionary/decode.cu [hunk 2]
```diff
@@ -30,6 +29,18 @@
 namespace cudf {
 namespace dictionary {
 namespace detail {
+namespace {
+struct indices_handler_fn {
```
Comment (mythrocks) [inline cpp/src/dictionary/decode.cu:33]: Thanks for fixing this.  Would there be value in naming this functor a "nullifying_handler" or some such?

**Review by mythrocks — APPROVED** [review]
LGTM!

File: cpp/src/dictionary/decode.cu (hunk 2 above)
Comment (davidwendt) [reply cpp/src/dictionary/decode.cu:33]: The actually nullifying happens far away in the gather function. This function simply returns indices.

**Pull request merged**